### ISSUE TITLE
String ordering produce wrong results when called from webassembly.

### ISSUE THREAD

**Issue by BaherBarakat100** [WebAssembly/binaryen#4034]
We noticed that the results of using the compare function in the std::collate is different when called using webassembly than when called using C++ native. 

The main bug, was that in webassembly, it add the strings which begin with capital letter to a different category, and sort it at the beginning of the list, while this is not the case in the C++ native code.
Other bug, was that the strings begin with German ümlaut characters, are also added to different category than the normal English characters, which also not the case in C++ native code. For example, when sorting using the C++ native, the 'a' should be treated like 'ä'.

Attached is screenshots of the difference between the sorting using C++ native code, and the sorting when called using webassembly.

Sorting using native code results:
![results-native](https://user-images.githubusercontent.com/88090999/127339117-e1944946-6d23-4e36-920c-169ed13968b9.PNG)
Sorting using webassembly:
![results-webassembly](https://user-images.githubusercontent.com/88090999/127337501-1ae7fd43-ac16-47e6-b891-c283ccc1ea6b.PNG)

We implemented a reproducer code using C++, which compare different strings, and found that when we use the same code with the same locales in the webassembly, different comparison output result.

The idea of the reproducers is to show the difference between C++  native and webassembly in comparing strings with different locales, using the same code.
[reproducer_webassembly.zip](https://github.com/WebAssembly/binaryen/files/6892984/reproducer_webassembly.zip)

[reproducer_native.zip](https://github.com/WebAssembly/binaryen/files/6894040/reproducer_native.zip)

You can do the following steps to execute the reproducers code using For webassembly:

Extract the zip folder, then execute test.html file.

1- Install and activate emscripten.

2- Run the shell script (compileandrun.sh)
 
Follow these steps, in order to show these results in webassembly:

Run the test.html file.
Click on compare button to show the comparison of the previously mentioned strings.
Click on the button related to each locale, in order to sort the strings according to this locale.
Click on the transform button related to each locale, in order to sort these strings according to this locale, but after transforming the strings to lower case.
to reproduce the previously discussed bugs.

For the native C++ reproducer:

1- Open the reproduce.sln file using visual studio.
2- Build the reproducer.
3- Start debugging.
 

In our reproducers we compared the following pairs of strings:

Anton, anton

anton, Anton

anton, änton

anton, Barbara

anton, Barbara

ßab, Öab

anton, ßab

 

with the following locales:

Empty Locale.
C Locale.
English Locale.
German Locale.
 

From the previous results, it was obvious that there were difference between the results of the C++ Native and webassembly.

**Comment by MaxGraey** [comment]
Does it reproduced with `-O0` (without optimization)?

**Comment by BaherBarakat100** [comment]
Hello MaxGraey, 

Thank you for your reply.

Yes, we reproduce with -O0 (without optimizations), and compile the native builds with MSVC compiler.

Thank you.

**Comment by MaxGraey** [comment]
> Yes, we reproduce with -O0 (without optimizations), and compile the native builds with MSVC compiler.

Could it be that this is just different behaviour of the standard library of Clang and MSVC? I remember several instances last year when the MSVC version of the library didn't work as well as GCC or Clang. Perhaps the problem has nothing to do with Binaryen or emscripten

**Comment by kripken** [comment]
It could also be an issue with how UTF data is sorted in the standard library. For wasm, it could be a musl libc bug.

I'd suggest simplifying the reproducer as much as possible. A single cpp file that can compile and run either natively or to wasm would be easiest in order to debug it.

**Comment by MaxGraey** [comment]
Right, it definitely musl issue due to it not implement most of locale parts:
https://wiki.musl-libc.org/open-issues.html

> Locale support is very limited, and barely works. Translation of LC_TIME is not possible because the key strings for ABMON_5 and MON_5 (“May”) are identical. Custom collation orders (LC_COLLATE) are not implemented at all, despite there always having been an intent to support them. LC_NUMERIC and LC_MONETARY also admit no variation by locale. Solving these problems requires a major overhaul, but the main missing prerequisite is involvement from users who want the functionality.

**Comment by BaherBarakat100** [comment]
![MSVC](https://user-images.githubusercontent.com/88090999/128194775-56360ee0-98ec-4268-8d2c-10d4c8730a59.png)
![MSVC](https://user-images.githubusercontent.com/88090999/128194834-4034cad2-57d0-4fe9-8dc1-376d304e385a.png)
> > Yes, we reproduce with -O0 (without optimizations), and compile the native builds with MSVC compiler.
> 
> Could it be that this is just different behaviour of the standard library of Clang and MSVC? I remember several instances last year when the MSVC version of the library didn't work as well as GCC or Clang. It's hard to say if this is a Clang or MSVC problem.

I used this online compiler: https://rextester.com/l/cpp_online_compiler_gcc, in order to try to compile my code using clang or gcc, and it had the same comparsion results of MSVC which ara different than web assembly results.
![compile-gcc](https://user-images.githubusercontent.com/88090999/128194761-74191e8c-b3b9-4b64-b0db-770e4ba1e211.PNG)
![clang-compile](https://user-images.githubusercontent.com/88090999/128195112-ba3f77d5-f002-4f7d-929f-8c18b901e700.PNG)
![MSVC](https://user-images.githubusercontent.com/88090999/128195129-8dd7408f-aa8e-4850-bcb0-4f31fb90efc4.png)

**Comment by BaherBarakat100** [comment]
> @BaherBarakat100 See my prev comment: [#4034 (comment)](https://github.com/WebAssembly/binaryen/issues/4034#issuecomment-890036040). So it seems this problem of musl library which used by emscripten

Thank you for your reply. Please, do you think, I should create an issue there, and do you think that there any future plans for  adding the locale support?

**Comment by MaxGraey** [comment]
Perhaps better open this issue in emscripten-core/emscripten. Although musl doesn't have this functionality as such and I'm not sure if emscripten can work with another standard library. Maybe the best solution would be to bypass this and use for example boost library or something else for this

**Comment by tlively** [comment]
Closing this because it is not a binaryen bug.
